engines/dfc_provider/app/controllers/dfc_provider/addresses_controller.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+# Controller used to provide the CatalogItem API for the DFC application
+module DfcProvider
+  class AddressesController < DfcProvider::ApplicationController
+    def show
+      address = Spree::Address.find(params.require(:id))
+
+      return not_found unless authorized(address)
+
+      dfc_address = AddressBuilder.address(address)
+      render json: DfcIo.export(dfc_address)
+    end
+
+    private
+
+    # Does the current user have access to this address?
+    #
+    # It's possible to guess address ids and therefore we need to prevent the
+    # collection of sensitive customer data.
+    #
+    # We may want to extend the list of authorised addresses once the DFC API
+    # references them. To start with, we would only need enterprise addresses
+    # but I included a few more options to show how this will probably evolve.
+    #
+    # Currently not checked models:
+    #
+    # - Spree::Card
+    # - Spree::Order
+    # - Spree::Shipment
+    # - Subscription
+    def authorized(address)
+      current_user.ship_address_id == address.id ||
+        current_user.bill_address_id == address.id ||
+        [
+          customer_address(address),
+          public_enterprise_group_address(address),
+          public_enterprise_address(address),
+          managed_enterprise_address(address),
+        ].any?(&:exists?)
+    end
```

Comment:
Yes, we could do that. At the moment, the main app doesn't have any authorisation logic around addresses because it only looks at the associated object. While we haven't implemented a complete solution yet and this isn't needed anywhere else in the app, I would like to keep it here, in a private method, clearly not affecting anything else.